### ISSUE TITLE
[Question] derivative with respect to ts for SaveAt snapshots

### ISSUE THREAD

**Issue by ASKabalan** [patrick-kidger/diffrax#584]
Hello 👋 

this is (sort of) a follow up for this discussion https://github.com/patrick-kidger/diffrax/issues/541

So I am in the process of making my own reverse adjoint for actual reversible solvers (not covered by https://github.com/patrick-kidger/diffrax/pull/528) and here is [my implementation](https://github.com/ASKabalan/JaxPM-Paper/blob/main/content/tools/integrate.py) (not AbstractAdjoint Compatible yet)

I am using both diffrax and vanila lax scan as reference and I noticed something weird

If the `ts` in `SaveAt.subs.ts` are in the autodiff graph (which is my use case .. one leaf of arg can influence the exact time at which we want to snapshot) I am getting different results between diffrax and just a  lax scan

In the JAX case, the gradients of an argument that influence the ts are always zeros
However the gradients with respect to y0 and args can change if this parameter changes and it matches diffrax result (see example)

Now this is when compared to `RecursiveCheckpointAdjoint`

With a `BackSolveAdjoint` (which per your recommendation is where I am getting inspiration) 
I am getting error 
`Unexpected tangent. `*args` and `**kwargs` to `filter_custom_vjp` cannot be autodifferentiated.`

Understandable, Because `SaveAt` is not the first argument in the `filter_custom_vjp`

Now in my case [I did this ](https://github.com/ASKabalan/JaxPM-Paper/blob/main/content/tools/integrate.py#L288) but I assume it is useless since it is taking adjoints with respect to the save_y function and not the entire step

In the example you notice that gradients with respect of y0 and args are correct but not x 
In my case x would be inside args so this is an issue

Do you have any mathematical insight on how should this happen?

My intuition is that changing the snapshot times should influence the arguments and y0 and there is no gradient to propagate so gradient should be zero.

Should I propagate the gradients of the function ODETerm.vfprod with respect of ts aswell ?

Then it should only be done only with the snapshot times correct?

Example 

```py

import jax
import jax.numpy as jnp
from diffrax import ODETerm , diffeqsolve , SaveAt , RecursiveCheckpointAdjoint , Euler , BacksolveAdjoint, SemiImplicitEuler
import os
import sys

def check_tree(x , y):
    return jax.tree.all(jax.tree.map(lambda x , y : jnp.allclose(x , y) , x , y))

def f(t, x, z):
    return x + z[0] + t * z[1]


def g(t, y, z):
    return y + z[0] + t * z[1]

def fg(t , x_y , z):
    x , y = x_y
    out =  f(t , x , z) , g(t , y , z)
    return jnp.stack(out , axis=0)

def diffrax_integrate(ode_terms , solver , t0 , t1 , dt0 , y0 , args, saveat , checkpoints=10):
    sol = diffeqsolve(ode_terms , solver , t0 , t1 , dt0 , y0 , args , saveat=saveat , adjoint=RecursiveCheckpointAdjoint(checkpoints=checkpoints))
    return sol.ys

# VANILLA JAX LAX SCAN
def scan_integrate(
    terms,
    solver,
    t0 ,
    t1 ,
    dt0,
    y0,
    args,
    saveat,
) :

  def forward_step(carry, t):
      y, args = carry
      t_next = t + dt0
      # Perform a single integration step
      y_next, _, _, _, _ = solver.step(
          terms, t, t_next, y, args, solver_state=None, made_jump=False
      )
      # Store current state and time for backward pass
      return (y_next, args), y_next


  # Prepare the main scan
  init_carry = (y0, args)
  t_steps = jnp.arange(t0, t1, dt0)

  # Perform the integration using jax.lax.scan
  (_, _), ys = jax.lax.scan(forward_step, init_carry, t_steps)

  # Prepend the initial state to the scanned states
  ys = jax.tree.map(lambda y0 , ys :jnp.concatenate((jnp.asarray(y0)[None , ...] , ys) , axis=0) , y0 , ys)


  # Extract the requested snapshot times from the integrator results
  snapshots = saveat.subs.ts
  snapshots = ((snapshots - t0) / dt0).astype(jnp.int32)
  return jax.tree.map(lambda x : x[snapshots] , ys)


y0 = (1.0, 1.0)
args = (2.0 , 5)
x = 2.0

def symp_diffrax(y0 , args , x):

  t0 = 0.0
  t1 = 10.0
  dt0 = 1
  ode_terms = (ODETerm(g) , ODETerm(f))
  solver = SemiImplicitEuler()
  saveat = SaveAt(ts=[t0  ,  t0 +  x * dt0])
  return diffrax_integrate(ode_terms , solver , t0 , t1 , dt0 , y0 , args , saveat=saveat , checkpoints=10)


def euler_diffrax(y0 , args , x):

  y0 = jnp.stack(y0 , axis=0)
  t0 = 0.0
  t1 = 10.0
  dt0 = 1
  ode_term = ODETerm(fg)
  saveat = SaveAt(ts=[t0  , t0 + x *  dt0])

  return diffeqsolve(ode_term , Euler() , t0 , t1 , dt0 , y0 , args , saveat=saveat , adjoint=BacksolveAdjoint(solver=Euler())).ys


def symp_jax(y0 , args , x):
  
  t0 = 0.0
  t1 = 10.0
  dt0 = 1
  ode_terms = (ODETerm(g) , ODETerm(f))
  solver = SemiImplicitEuler()
  saveat = SaveAt(ts=[t0  , t0 +  x * dt0])
  return scan_integrate(ode_terms , solver , t0 , t1 , dt0 , y0 , args , saveat=saveat )


def euler_jax(y0 , args , x):

  y0 = jnp.stack(y0 , axis=0)
  t0 = 0.0
  t1 = 10.0
  dt0 = 1
  ode_term = ODETerm(fg)
  saveat = SaveAt(ts=[t0  , t0 + x *  dt0])

  return scan_integrate(ode_term , Euler() , t0 , t1 , dt0 , y0 , args , saveat=saveat)


def symp_my(y0 , args , x):
  
  t0 = 0.0
  t1 = 10.0
  dt0 = 1
  ode_terms = (ODETerm(g) , ODETerm(f))
  solver = SemiImplicitEuler()
  saveat = SaveAt(ts=[t0  , t0 +  x * dt0])
  return integrate(ode_terms , solver , t0 , t1 , dt0 , y0 , args , saveat=saveat )

sym_diffrax_grad_y0 = jax.jacrev(symp_diffrax , argnums=(0 , 1) , allow_int=True)(y0 , args , x)
sym_jax_grad_y0 = jax.jacrev(symp_jax , argnums=(0 , 1) , allow_int=True)(y0 , args , x)


sym_diffrax_grad_y0_2 = jax.jacrev(symp_diffrax , argnums=(0 , 1) , allow_int=True)(y0 , args , 2 * x)
sym_jax_grad_y0_2 = jax.jacrev(symp_jax , argnums=(0 , 1) , allow_int=True)(y0 , args , 2 * x)

sym_my_grad_y0 = jax.jacrev(symp_my , argnums=(0 , 1) , allow_int=True)(y0 , args , x)
sym_my_grad_y0_2 = jax.jacrev(symp_my , argnums=(0 , 1) , allow_int=True)(y0 , args , 2 * x)

print(f"Y0 gradient with x is same {check_tree(sym_diffrax_grad_y0 , sym_jax_grad_y0)}")  # True
print(f"Y0 gradient with 2x is same {check_tree(sym_diffrax_grad_y0_2 , sym_jax_grad_y0_2)}")  # True
print(f"Y0 gradient with x is same {check_tree(sym_diffrax_grad_y0 , sym_my_grad_y0)}")  # True
print(f"Y0 gradient with 2x is same {check_tree(sym_diffrax_grad_y0_2 , sym_my_grad_y0_2)}")  # True


sym_diffrax_grad_ts = jax.jacrev(symp_diffrax , argnums=2 , allow_int=True)(y0 , args , x)
sym_jax_grad_ts = jax.jacrev(symp_jax , argnums=2 , allow_int=True)(y0 , args , x)


sym_diffrax_grad_ts_2 = jax.jacrev(symp_diffrax , argnums=2 , allow_int=True)(y0 , args , 2 * x)
sym_jax_grad_ts_2 = jax.jacrev(symp_jax , argnums=2 , allow_int=True)(y0 , args , 2 * x)

sym_my_grad_ts = jax.jacrev(symp_my , argnums=2 , allow_int=True)(y0 , args , x)
sym_my_grad_ts_2 = jax.jacrev(symp_my , argnums=2 , allow_int=True)(y0 , args , 2 * x)


print(f"Y0 gradient with x is same {check_tree(sym_jax_grad_ts , sym_diffrax_grad_ts)}") #False
print(f"Y0 gradient with 2x is same {check_tree(sym_jax_grad_ts_2 , sym_diffrax_grad_ts_2)}") #False
print(f"Y0 gradient with x is same {check_tree(sym_jax_grad_ts , sym_my_grad_ts)}") # True (is zeros)
print(f"Y0 gradient with 2x is same {check_tree(sym_jax_grad_ts_2 , sym_my_grad_ts_2)}") # True (is zeros)
```

**Comment by patrick-kidger** [comment]
Hmm, sorry -- your example is quite large and appears to cover multiple questions?

FWIW to address at least this point:

> In the JAX case, the gradients of an argument that influence the ts are always zeros

here is a MWE demonstrating gradients with respect to `SaveAt(ts=...)`:

```python
import diffrax as dfx
import jax
import jax.numpy as jnp

@jax.jit
@jax.grad
def f(ts):
    sol = dfx.diffeqsolve(dfx.ODETerm(lambda t, y, args: -y), dfx.Tsit5(), t0=0, t1=1, dt0=0.1, y0=1, saveat=dfx.SaveAt(ts=ts))
    return jnp.sum(sol.ys)

grads = f(jnp.array([0.3, 0.5]))
print(grads)  # [-0.74081725 -0.6065303 ]
```

Perhaps you can break things down into a smaller MWE that demonstrates just a single thing, like the example I have here?

**Comment by ASKabalan** [comment]
Hello,

Sorry for the lengthy example.

From your example I can show that BacksolveAdjoint does not support differentiating w.r.t ts
```py
import diffrax as dfx
import jax
import jax.numpy as jnp

@jax.jit
@jax.grad
def f(ts):
    sol = dfx.diffeqsolve(dfx.ODETerm(lambda t, y, args: -y), dfx.Tsit5(), t0=0, t1=1, dt0=0.1, y0=1, saveat=dfx.SaveAt(ts=ts) , adjoint=dfx.BacksolveAdjoint(solver=dfx.Tsit5()))
    return jnp.sum(sol.ys)

grads = f(jnp.array([0.3, 0.5]))
print(grads)  # [NOT DIFFERENTIABLE]
```

Because ts is not part of the first argument in here 
https://github.com/patrick-kidger/diffrax/blob/14baa1edddcacf27c0483962b3c9cf2e86e6e5b6/diffrax/_adjoint.py#L537

I have kind of figured out a way to do with my own custom_vjp 
My idea is to propagate the gradients w.r.t of `t` in `f(t , y , args)` at snapshot times and return a list of gradients with same size as `ts` along side `diff_y`, `diff_arg` and `diff_term` as you did here

https://github.com/patrick-kidger/diffrax/blob/14baa1edddcacf27c0483962b3c9cf2e86e6e5b6/diffrax/_adjoint.py#L732

I would highly appreciate your feedback on this idea

**Comment by patrick-kidger** [comment]
Ah, right! Indeed, `BacksolveAdjoint` does not support this. I don't think this is a feature I expect to add to Diffrax -- essentially because `BacksolveAdjoint` should essentially never be used as it's such a bad algorithm, so there are simply more impactful features to spend my time implementing/reviewing.

As for your own in implementing a custom reversible adjoint, I can see that you might reasonably want gradients with respect to `SaveAt(ts=...)`. I think in order to make that happen you'd need to adjust `AdjointTerm` (i.e. write your own version of this) that also integrates the appropriate derivatives. (I'm willing to be wrong about though, been a while since I've thought about that detail ;) )

**Comment by ASKabalan** [comment]
Thank you

I am aware that it is not the algorithm to use .. I am just using it for inspiration.
But the idea of propagating the gradients w.r.t to ts seems to be working and returning the same resutls as `RecursiveCheckpointAdjoint`

Thank you again for your help
